Referentially separate backoff functions

kriskowal opened the pull request:
Without this change, it’s possible for backoff strategies to share state through a common random number generator, which leads to data races.

This change changes a backoff strategy to a constructor for lockless but referentially independent backoff functions. The random number generator option is now a constructor function.

akshayjshah commented:
Glad that we discovered this race now! I agree with Peter that closures feel like an unusual way to maintain private state in Go.

Would this feel more idiomatic if had something like this?

```
type Strategy interface {
  Backoff(attempts uint) time.Duration
}

// optionally, we could also have
type BackoffFunc func(uint) time.Duration
func (f BackoffFunc) Backoff(n uint) time.Duration { return f(n) }

func NewExponential(opts ...ExponentialOption) (Strategy, error) { ... }
```

akshayjshah commented on internal/backoff/exponential.go:
```diff
@@ -35,8 +35,7 @@ type ExponentialOption func(*exponentialOptions)
 // exponentialOptions are the configuration options for an exponential backoff
 type exponentialOptions struct {
 	base, min, max time.Duration
```
Under what circumstances would I want a non-zero minimum? As long as we're jittering, isn't zero safe for the server and fastest for the client?

akshayjshah replied:
Also...consider whether we should choose a new name for `base`. (Perhaps `initial`?)

In the context of exponentiation, I expected `base` to be `M` in `M ^ n`. That made me immediately suspicious of the `time.Duration` type.

willhug replied:
whoops, good point.

abhinav commented on internal/backoff/exponential.go:
```diff
@@ -55,12 +54,45 @@ func (e exponentialOptions) validate() (err error) {
+// Exponential strategies are not thread safe.
+// Use the Isolate() method to obtain an independent backoff generator with the
+// same options and a referentially independent random number generator.
+var DefaultExponential = (&exponentialStrategy{
```
If this is a `backoff.Strategy`, annotate the var with the type. Otherewise the type of `DefaultExponential` will be `func(uint) Duration`.

    var DefaultExponential Strategy = ...

abhinav commented on internal/backoff/exponential.go:
```diff
@@ -84,24 +116,28 @@ func MinBackoff(t time.Duration) ExponentialOption {
+//
+// Backoff functions are lockless and referentially independent, but not
+// thread-safe.
+func NewExponential(opts ...ExponentialOption) (func() func(uint) time.Duration, error) {
```
Shouldn't this return a Strategy?

kriskowal commented:
Is everyone in favor of adding a backoff.Strategy interface to `yarpc/api/backoff`? It’ll have to be moved out of internal since it’s an argument to public ConnBackoff transport options.

akshayjshah commented:
I'd prefer to add a strategy interface instead of committing to the function type in a public API. I can be convinced, though :)

kriskowal commented:
I’ve posted changes to address feedback and incidentally expand the scope of the change.

I’ve created an api/backoff with Strategy and Backoff interfaces. These are threaded through later changes to the options of TChannel and HTTP transport ConnBackoff constructors.

I’ve also changed the implementation to use only (first, max) parameters, removing (min) and renaming (base). These changes I’ve threaded through the config change following this one.

I’ve renamed the "Short" strategy and confined its type to the Strategy interface since it isn’t necessarily implemented as a degenerate ExponentialStrategy. It’s effectively a Noop type and could be implemented as a singleton in a future pass, but I’ve left it as is for expedience.

I think that I might send a note back in time suggesting that all API packages should have an "api" suffix like "peerapi" and "backoffapi", since it’s become pretty common to have to import both the API and implementation packages.

peter-edge commented on internal/backoff/exponential.go:
```diff
@@ -34,37 +35,47 @@ type ExponentialOption func(*exponentialOptions)
-	base, min, max time.Duration
-	rand           *rand.Rand
-	minMaxDiff     int64
+	first, max time.Duration
```
nit: I don't like this style of declaring multiple variables with the same type in the same line, and we don't do it in other places in the codebase I don't think

peter-edge commented on internal/backoff/exponential.go:
```diff
@@ -34,37 +35,47 @@ type ExponentialOption func(*exponentialOptions)
-	if e.min < 0 {
-		err = multierr.Append(err, errors.New("invalid min for exponential backoff, need greater than or equal to zero"))
+	if e.first <= 0 {
+		err = multierr.Append(err, errors.New("invalid first duration for exponential backoff, need greater than zero"))
```
nit (but not really part of this PR): I like most errors created with `errors.New` to be global constants.

peter-edge commented on internal/backoff/exponential.go:
```diff
@@ -34,37 +35,47 @@ type ExponentialOption func(*exponentialOptions)
-	base: time.Millisecond,
-	max:  time.Hour, // :shrug:
-	rand: rand.New(rand.NewSource(time.Now().UnixNano())),
+	first:   100 * time.Millisecond,
```
Can we make `defaultExponentialOpts` a public variable so this can be seen in documentation?  We should also talk about all the arbitrary default durations around the codebase.

willhug commented on internal/backoff/exponential.go:
```diff
@@ -26,6 +26,7 @@ import (
 	"time"
 
 	"go.uber.org/multierr"
+	"go.uber.org/yarpc/api/backoff"
```
ordering?

willhug commented on internal/backoff/exponential_test.go:
```diff
@@ -39,57 +46,16 @@ func TestExponential(t *testing.T) {
-				"invalid min for exponential backoff, need greater than or equal to zero",
-			},
-		},
-		{
```
we can keep the invalid max test
